### PR TITLE
Code first UI

### PR DESCRIPTION
**Pull request by AnDeVerin**

### CODE REVIEW

File: backend/.eslintrc.cjs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  extends: ['eslint:recommended', 'plugin:@typescript-eslint/recommended'],
+  parser: '@typescript-eslint/parser',
+  parserOptions: {
+    ecmaVersion: 'latest',
+    sourceType: 'module',
+  },
+  plugins: ['@typescript-eslint', 'simple-import-sort'],
+  rules: {
+    'simple-import-sort/imports': 'error',
+    'simple-import-sort/exports': 'error',
```
Comment (embryCODE) [inline backend/.eslintrc.cjs:15]: I added this to mine too

File: backend/src/utils/ResponseDTO.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+export class ResponseDTO<T> {
```
Comment (embryCODE) [inline backend/src/utils/ResponseDTO.ts:1]: I did this response DTO thing too! nice

File: backend/src/models/userModel.ts [hunk 3]
```diff
@@ -47,7 +47,7 @@ export const createUser = async ({
     });
 
     return newUser;
-  } catch (error) {
+  } catch {
     throw new Error();
   }
```
Comment (embryCODE) [inline backend/src/models/userModel.ts:52]: This is what we call a "useless catch." To catch and rethrow without doing anything means that you should not have this try/catch at all. In general, only catch when you can recover from the error gracefully with some kind of fallback, otherwise, just let it crash. But a catch and rethrow is like doing nothing at all.

File: backend/src/models/userModel.ts [hunk 4]
```diff
@@ -102,7 +102,7 @@ export const editUser = async ({
       firstName: updatedUser.firstName,
       lastName: updatedUser.lastName,
     };
-  } catch (error) {
+  } catch {
     throw new Error();
   }
```
Comment (embryCODE) [inline backend/src/models/userModel.ts:107]: I just realized you might be doing it to hide details of the thrown errors. I guess that could make sense, although you can do that at the layer where you handle the thrown errors.

File: frontend/.eslintrc.cjs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    'plugin:@tanstack/eslint-plugin-query/recommended',
+  ],
+  ignorePatterns: ['dist', '.eslintrc.cjs'],
+  parser: '@typescript-eslint/parser',
+  plugins: ['react-refresh', 'simple-import-sort'],
+  rules: {
+    'react-refresh/only-export-components': [
+      'warn',
+      { allowConstantExport: true },
+    ],
```
Comment (embryCODE) [inline frontend/.eslintrc.cjs:17]: I ran into this myself and dropped in an ignore.

File: frontend/src/components/App/App.tsx [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import 'react-toastify/dist/ReactToastify.css';
+
+import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
```
Comment (embryCODE) [inline frontend/src/components/App/App.tsx:3]: react-query, nice. I may end up using that. I've used rtk-query before, which is similar but with redux.

File: frontend/src/components/PostCard/PostCard.tsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import Link from 'antd/es/typography/Link';
+import Title from 'antd/es/typography/Title';
+import { differenceInCalendarDays } from 'date-fns';
+
+import { Post, Vote, VoteType } from '@/utils/types';
+
+import styles from './PostCard.module.css';
+
+const getPostAge = (postDate: Date) => {
+  const diffInDays = differenceInCalendarDays(new Date(), postDate);
```
Comment (embryCODE) [inline frontend/src/components/PostCard/PostCard.tsx:12]: I was lazy and used dayjs for this

File: frontend/src/components/PostCard/PostCard.tsx [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  return diffInDays ? `${diffInDays} days ago` : 'Today';
+};
+
+const getRating = (votes: Vote[]) => {
+  return votes.reduce((acc, vote) => {
+    return acc + (vote.voteType === VoteType.UPVOTE ? 1 : -1);
+  }, 0);
```
Comment (embryCODE) [inline frontend/src/components/PostCard/PostCard.tsx:19]: I just realized I forgot to do this in my PR. Thank you!

File: frontend/src/components/PostCard/PostCard.tsx [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+};
+
+export const PostCard = (props: Post) => {
```
Comment (embryCODE) [inline frontend/src/components/PostCard/PostCard.tsx:22]: So, I'm sure Khalil will walk us through this but I took a slightly different strategy. I have a `PostViewModel` type which is just simple primitive values like strings and numbers. Then in a layer outside of the UI components I'm converting from `Post` to `PostViewModel`. Just an idea.

File: frontend/src/pages/MainPage/MainPage.tsx [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    ],
+    [tabContent]
+  );
+
+  return (
+    <Tabs
+      activeKey={activeTabKey}
+      items={tabItems}
+      size="large"
+      onChange={(activeKey) => setActiveTabKey(activeKey as ActiveTabKeys)}
```
Comment (embryCODE) [inline frontend/src/pages/MainPage/MainPage.tsx:52]: Why did you need the type assertion here? These make me nervous.

File: frontend/src/pages/RegistrationPage/RegistrationPage.tsx [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+      span: 16,
+      offset: 8,
+    },
+  },
+};
+
+const onFinish = async (
+  values: UserRegistrationData
+): Promise<UserRegistrationResponse> => {
+  const responce = await fetch(`${HOST}/users/new`, {
```
Comment (embryCODE) [inline frontend/src/pages/RegistrationPage/RegistrationPage.tsx:45]:
```suggestion
  const response = await fetch(`${HOST}/users/new`, {
```

File: frontend/src/pages/RegistrationPage/RegistrationPage.tsx [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+      offset: 0,
+    },
+    sm: {
+      span: 16,
+      offset: 8,
+    },
+  },
+};
+
+const onFinish = async (
```
Comment (embryCODE) [inline frontend/src/pages/RegistrationPage/RegistrationPage.tsx:42]: "onFinish" is slightly confusing. I think I'd call this `createUser` or something like that. Or `createUserAPICall`

File: frontend/src/pages/RegistrationPage/RegistrationPage.tsx [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+      setIsLoading(true);
+      const { success, data, error } = await onFinish(values);
+
+      if (success && data) {
+        toast('Success! Redirecting home.', {
+          autoClose: 3000,
+          type: 'success',
+        });
+        setUser(data); // setUser is a function that sets the user in the context (App.tsx
+        setTimeout(() => navigate('/'), 3000);
```
Comment (embryCODE) [inline frontend/src/pages/RegistrationPage/RegistrationPage.tsx:70]: I did a timeout for this too, to give people time to read the toast message.

File: frontend/src/pages/RegistrationPage/RegistrationPage.tsx [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+      } else {
+        const errorMessage = error ? ErrorTexts[error] : 'Unknown error';
+        toast(errorMessage, { type: 'error' });
+        setIsLoading(false);
```
Comment (embryCODE) [inline frontend/src/pages/RegistrationPage/RegistrationPage.tsx:74]: You can put `setIsLoading` in a `finally` block so it stops loading whether success or failure. (Of course you redirect on success, so maybe not necessary.

File: frontend/src/utils/api/usePostsQuery.ts [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+): Promise<Post[]> => {
+  const queryParamsAsStrings = Object.entries(queryParams).map(
+    ([key, value]) => [key, value.toString()]
+  );
+
+  const querystring = new URLSearchParams([...queryParamsAsStrings]).toString();
+
+  const response = await fetch(`${HOST}/posts?${querystring}`);
+  const { success, data } = await response.json();
+  return success ? data : [];
```
Comment (embryCODE) [inline frontend/src/utils/api/usePostsQuery.ts:23]: If not success, are you doing anything to indicate that in the UI? (I just realized I forgot to handle this error in my code too...)

File: frontend/src/utils/constants/errors.ts [hunk 16]
```diff
@@ -0,0 +1,14 @@
+export enum ErrorMessages {
+  VALIDATION_ERROR = 'ValidationError',
+  USER_NOT_FOUND = 'UserNotFound',
+  EMAIL_IN_USE = 'EmailAlreadyInUse',
+  USERNAME_IN_USE = 'UsernameAlreadyTaken',
+}
+// Copy from backend/src/utils/errors/constants.ts
```
Comment (embryCODE) [inline frontend/src/utils/constants/errors.ts:7]: I wonder if we will eventually have a shared dir between frontend and backend?

File: frontend/src/utils/hooks/useUserSetter.ts [hunk 17]
```diff
@@ -0,0 +1,7 @@
+import { useOutletContext } from 'react-router-dom';
+
+import { UserSetterContextType } from '../types/user';
+
+export function useUserSetter() {
+  return useOutletContext<UserSetterContextType>();
```
Comment (embryCODE) [inline frontend/src/utils/hooks/useUserSetter.ts:6]: Hmm, I've never heard of this

File: frontend/src/utils/types/vote.ts [hunk 18]
```diff
@@ -0,0 +1,11 @@
+export enum VoteType {
+  UPVOTE = 'UPVOTE',
+  DOWNVOTE = 'DOWNVOTE',
+}
```
Comment (embryCODE) [inline frontend/src/utils/types/vote.ts:4]: Nice. I saw what you did in Prisma for this. I used `'Upvote' | 'Downvote'` union type for mine, but an enum is probably better.

**Review by embryCODE — APPROVED** [review]
Nice work. This helped me remember some things I forgot in my code!

File: backend/src/models/userModel.ts (hunk 4 above)
Comment (AnDeVerin) [reply backend/src/models/userModel.ts:107]: You are right. I am masking the initial error with an empty error here. Probably, it's better to do that at the error handler layer.

File: frontend/src/components/PostCard/PostCard.tsx (hunk 9 above)
Comment (AnDeVerin) [reply frontend/src/components/PostCard/PostCard.tsx:22]: I agree. There must be a better solution. I decided to make it simple for this iteration and introduce an adapter later.

File: frontend/src/pages/MainPage/MainPage.tsx (hunk 10 above)
Comment (AnDeVerin) [reply frontend/src/pages/MainPage/MainPage.tsx:52]: `onChange` returns string, so I have to assert the type explicitly here

File: frontend/src/utils/api/usePostsQuery.ts (hunk 15 above)
Comment (AnDeVerin) [reply frontend/src/utils/api/usePostsQuery.ts:23]: Not really. I'll do it later on in the course.

File: frontend/src/utils/constants/errors.ts (hunk 16 above)
Comment (AnDeVerin) [reply frontend/src/utils/constants/errors.ts:7]: I hope so )

**Comment by AnDeVerin** [comment]
@embryCODE thank you for your review. )
